src/router.rs

```diff
@@ -190,18 +207,26 @@ impl Router {
             url = Url::from_generic_url(generic_url).unwrap();
         }
 
-        self.recognize(&req.method, &path).and(
+        self.recognize(&req.method, &path).ok().and(
             Some(IronError::new(TrailingSlash,
                                 (status::MovedPermanently, Redirect(url))))
         )
     }
 
     fn handle_method(&self, req: &mut Request, path: &str) -> Option<IronResult<Response>> {
-        if let Some(matched) = self.recognize(&req.method, path) {
-            req.extensions.insert::<Router>(matched.params);
-            req.extensions.insert::<RouterInner>(self.inner.clone());
-            Some(matched.handler.handle(req))
-        } else { self.redirect_slash(req).and_then(|redirect| Some(Err(redirect))) }
+        match self.recognize(&req.method, path) {
+            Ok(matched) => {
+                req.extensions.insert::<Router>(matched.params);
+                req.extensions.insert::<RouterInner>(self.inner.clone());
+                Some(matched.handler.handle(req))
+            },
+            Err(e) => {
+                if format!("{:?}", e.error) == "MethodNotAllowed" {
```

Comment:
This is what I'm referring to in a previous comment. Not returning IronError from recognize (but a concrete error type) fixes that.